cpp/src/svm/smosolver.h

```diff
@@ -0,0 +1,271 @@
+ * Sequential Minimal Optimization (SMO).
+ *
+ * The general decomposition idea by Osuna is to choose q examples from all the
+ * training examples, and solve the QP problem for this subset (discussed in
+ * section 11.2 by Joachims [1]). SMO is the extreme case where we choose q=2.
+ *
+ * Here we follow [2] and [3] and use two level decomposition. First we set
+ * q_1=1024, and solve the QP sub-problem for that (let's call it QP1). This is
+ * the outer iteration, implemented in SmoSolver::Solve.
+ *
+ * To solve QP1, we use another decomposition, specifically the SMO (q_2 = 2),
+ * which is implemented in SmoBlockSolve.
+ *
+ * References:
+ * - [1] Joachims, T. Making large-scale support vector machine learning
+ *      practical. In B. Scholkopf, C. Burges, & A. Smola (Eds.), Advances in
+ *      kernel methods: Support vector machines. Cambridge, MA: MIT Press (1998)
+ * - [2] J. Vanek et al. A GPU-Architecture Optimized Hierarchical Decomposition
+ *      Algorithm for Support VectorMachine Training, IEEE Transactions on
+ *      Parallel and Distributed Systems, vol 28, no 12, 3330, (2017)
+ * - [3] Z. Wen et al. ThunderSVM: A Fast SVM Library on GPUs and CPUs, Journal
+ *      of Machine Learning Research, 19, 1-5 (2018)
+ */
+template <typename math_t>
+class SmoSolver {
+ public:
+  bool verbose = false;
+  SmoSolver(const cumlHandle_impl &handle, math_t C, math_t tol,
+            MLCommon::GramMatrix::GramMatrixBase<math_t> *kernel,
+            float cache_size = 200)
+    : handle(handle),
+      n_rows(n_rows),
+      C(C),
+      tol(tol),
+      kernel(kernel),
+      cache_size(cache_size),
+      stream(handle.getStream()),
+      return_buff(handle.getDeviceAllocator(), stream, 2),
+      alpha(handle.getDeviceAllocator(), stream),
+      delta_alpha(handle.getDeviceAllocator(), stream),
+      f(handle.getDeviceAllocator(), stream) {}
+
+#define SMO_WS_SIZE 1024
+  /**
+   * Solve the quadratic optimization problem.
+   *
+   * The output arrays (dual_coefs, x_support, idx) will be allocated on the
+   * device, they should be unallocated on entry.
+   *
+   * @param [in] x training vectors in column major format, size [n_rows x n_cols]
+   * @param [in] n_rows number of rows (training vectors)
+   * @param [in] n_cols number of columns (features)
+   * @param [in] y labels (values +/-1), size [n_rows]
+   * @param [out] dual_coefs, size [n_support] on exit
+   * @param [out] n_support number of support vectors
+   * @param [out] x_support support vectors in column major format, size [n_support, n_cols]
+   * @param [out] idx the original training set indices of the support vectors, size [n_support]
+   * @param [out] b scalar constant for the decision function
+   * @param [in] max_out_iter maximum number of outer iteration (default 100 * n_rows)
+   * @param [in] xm_inner_iter maximum number of inner iterations (default 10000)
+   */
+  void Solve(math_t *x, int n_rows, int n_cols, math_t *y, math_t **dual_coefs,
```

Comment:
We usually use "fit" rather than "solve" for these type of functions. To be same with other solvers, I'd rename this function to "fit".